File: coin_change.cpp
#include<iostream>
#include<vector>
#include<string>
#include<stack>
#include<algorithm>
#include<queue>
#include<math.h>

using namespace std;

int dp[4][5];

int count(int arr[], int m, int n) {
    if(n<0) return 0;
    if(n==0) return 1;
    if(m<=0 && n>=1) return 0;
    if(dp[m][n]) return dp[m][n];

    dp[m][n] = count(arr,m-1,n) + count(arr,m,n-arr[m-1]);
    return dp[m][n];
}

int main() {
    int i, j; 
    int arr[] = {1, 2, 3}; 
    int m = sizeof(arr)/sizeof(arr[0]); 
    cout<<count(arr, m, 4)<<endl;
    return 0;
}

File: coin_change.go
package main

import "fmt"

var dp [][]int

func getDpArray(m int, n int) [][]int {
	dp := make([][]int, m)
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n)
	}
	return dp
}

func count(arr []int, m int, n int) int {
	if n < 0 {
		return 0
	}
	if n == 0 {
		return 1
	}

	if m <= 0 && n >= 1 {
		return 0
	}

	if dp[m][n] != 0 {
		return dp[m][n]
	}

	dp[m][n] = count(arr, m-1, n) + count(arr, m, n-arr[m-1])
	return dp[m][n]
}

func main() {
	n := 4
	arr := []int{1, 2, 3}
	dp = getDpArray(4, 5)
	fmt.Println(count(arr, len(arr), n))
}
